contracts/DKVDaggerHashimoto.sol

```diff
@@ -277,32 +358,57 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
         uint256 startShardId,
         uint256 shardLenBits,
         bytes32 hash0,
+        address miner,
         bytes32[][] memory proofsDim2,
         bytes[] memory maskedData
     ) internal view returns (bytes32) {
-        require(maskedData.length == randomChecks, "data vs checks: length mismatch");
-        require(proofsDim2.length == randomChecks, "proofs vs checks: length mismatch");
-        uint256 maxKvSize = 1 << maxKvSizeBits;
+        require(
+            maskedData.length == randomChecks,
+            "data vs checks: length mismatch"
+        );
+        require(
+            proofsDim2.length == randomChecks,
+            "proofs vs checks: length mismatch"
+        );
+        // calculate the number of chunks range of the sample check
         uint256 rows = 1 << (shardEntryBits + shardLenBits + chunkLenBits);
 
         for (uint256 i = 0; i < randomChecks; i++) {
-            require(maskedData[i].length == chunkSize, "invalid proof size");
+            uint256 mChunkSize = chunkSize;
+            require(maskedData[i].length == mChunkSize, "invalid proof size");
             uint256 parent = uint256(hash0) % rows;
-            uint256 chunkIdx = parent + (startShardId << (shardEntryBits + chunkLenBits));
+            uint256 chunkIdx = parent +
+                (startShardId << (shardEntryBits + chunkLenBits));
             uint256 kvIdx = chunkIdx >> chunkLenBits;
-            chunkIdx = chunkIdx % (1 << chunkLenBits);
-            bytes memory data = systemContract.unmaskWithEthash(kvIdx, maskedData[i]);
+            PhyAddr memory kvInfo = kvMap[idxMap[kvIdx]];
+
+            /* NOTICE: kvInfo.hash is the bytes24-hash , the value will be set at the front
+             *          24-byte when it converts to bytes32-hash */
+            bytes memory unmaskedData = systemContract.unmaskChunkWithEthash(
+                uint64(chunkIdx),
+                kvInfo.hash,
+                miner,
+                maskedData[i]
+            );
             /* NOTICE: Now we have kvIdx and chunkIdx both generated from hash0
              *         The difficulty should increase intrinsically */
             require(
-                checkDaggerDataWithProof(chunkIdx, kvMap[idxMap[kvIdx]].hash, proofsDim2[i], data),
+                checkDaggerDataWithProof(
+                    chunkIdx,
+                    kvInfo,
+                    proofsDim2[i],
+                    unmaskedData
+                ),
                 "invalid access proof"
             );
 
+            /* NOTICE: we should use the maskedChunkData merged with the `hash0` to calculate the new `hash0`
+             *          because the miner executes this `hash0` calculation off-chain in this way. */
+            bytes memory maskedChunkData = maskedData[i];
```

Comment:
Optimization view: Is this copy worthy? Can we directly use maskedData[i] to compute next hash?

Reply:
the assembly can not use maskedData[i]